### ISSUE TITLE
confusion about scale factor

### ISSUE THREAD

**Issue by sunnyHelen** [ClementPinard/SfmLearner-Pytorch#59]
![image](https://user-images.githubusercontent.com/28101548/54660429-8a268280-4b2a-11e9-8fc8-a03ff9c2a10e.png)
What's the difference between the two ways for calculating errors?
I changed the model a little and trained 1000 epochs using your 'dispnet_model_best.pth.tar' and 'exp_pose_checkpoint.pth.tar' as pretrained model.
But I get worse results in 'Results with scale factor determined by GT/prediction ratio (like the original paper) ' but better result in 'Results with scale factor determined by PoseNet'. What's dose it mean? How many epochs will be better?

**Comment by ClementPinard** [comment]
it's interesting that you find better results with posenet !

Basically, posenet provides a way to compute scale factor not from median of ground truth of depth, but from displacement estimation from GPS in KITTI metadata.

As such, it reflects a much more practical usecase where this single frame depth would be used. I wrote a pargraph on this in this paper, section 3 http://perso.ensta-paristech.fr/~pinard/unsupervised-depthnet/pdf/learning-structure-motion.pdf (yes I'm doing my own publicity ! :stuck_out_tongue: )

**Comment by ClementPinard** [comment]
as for the epochs,

**Comment by sunnyHelen** [comment]
Wow, congratulations. You paper must be a really great paper. 
It seems in the case scale factor determined by GT, the evaluation quantative  error results tend to much lower. But in my mind the factor determined by poseNet is more meaningful. Am I right?
And also in my changed model, I found the VO result is much worse than the result you post. I'm still finding the reason. Do you have any suggestion?
Another question is the result you post is trained by KITTI only or also CS?
Thank you so much for kindly discussing with me~

**Comment by sunnyHelen** [comment]
![image](https://user-images.githubusercontent.com/28101548/54733476-003fed80-4bee-11e9-971c-4f57a0be7916.png)
And can training with gt or not  affect the result?

**Comment by ClementPinard** [comment]
no cityscape training. See a note from the original author : https://github.com/tinghuiz/SfMLearner#notes

As for the scale ratio determination, it depends on what you are interested in. You can only want to know the depth relative to other ares in the image. see a video from niantic that uses a depth network for VR with occlusion : https://www.youtube.com/watch?v=7ZrmPTPgY3I
In this case they used a depth network trained in the same fashion, but they don't want to know absolute depth, only occlusion areas with respect to a plane.

For navigation (which is kinda the point of KITTI), you might either take the scale factor from PoseNet + speed measurement, or you can do depth completion with a few GT values, if you imagine a very sparse  and light Lidar instead of the beast that is used by KITTI for its depth ground truth

**Comment by ClementPinard** [comment]
the training with gt will affect the saved model : the "model_best" will be the one that performs the best on the validation set. The key idea in adding this validation set that has GT is that author observed that training gets worse if the training has too many iterations. To know that, they used the test set, and checked that networks saved earlier in the training did perform better.

This is not good ! This is basically early stopping applied on test set, which is not truly unsupervised anymore, and very prone to overfitting. To address this issue, I implemented a proper early stopping mechanism with a validation set that is NOT the test set. It now two sets with GT instead of one, but it's less hypocritical regarding the term "unsupervised".

**Comment by ClementPinard** [comment]
And lastly, regarding the Visual Odometry results, Depthg network ususally peak before PoseNet. The early stopping mechanism explained above save both networks at the same time. So the resulting posenet is not very good, but if you take the best depthnet and the best posenet from two different epochs, they won't be consistent with each other, and the scale factor determined by PoseNet will have chances be very bad.

From personal experience, if you only want VO without depth, you can take the posenet_checkpoint (last version, regardless of validation score) instead of posenet_model_best .

Hope I was helpful,
Clément

**Comment by sunnyHelen** [comment]
Thank you for your suggestion. Sadly, my posenet_checkpoint result is very very bad too. And also the depth result with scale factor determined by GT/prediction ratio is also very bad. I'm still doing experiments to check different iteration numbers.

**Comment by sunnyHelen** [comment]
![image](https://user-images.githubusercontent.com/28101548/54863990-0081e980-4da4-11e9-8606-426493c0a810.png)
Sorry to bother you again~
I read your paper and for the quantative result from zhou et al using factor by Posenet, which model do you use? checkpoint or best?

**Comment by ClementPinard** [comment]
Factor from posent needs to be used with the same network as depthnet, so you need to do it either on both "best" networks or "checkpoint" network, but it has to be the same.

I probably used the "best" for both depthnet and posenet here

**Comment by sunnyHelen** [comment]
![image](https://user-images.githubusercontent.com/28101548/54895091-fbdb4380-4f11-11e9-855d-6e46eb6d38ef.png)
But the result I got using both "best", is like that. Result about scale factor determined by Posenet is very bad. 
I can't get the good result as yours in your paper.

**Comment by ClementPinard** [comment]
Is that your network or the pretrained I made available for download ? I recall having a problem with them recently, so maybe it's that. Reuploading a good "posenet_model_best" is in my todo list

**Comment by sunnyHelen** [comment]
It's the pretrained "best" you provided. I test it and find it's very different from your paper.

**Comment by ClementPinard** [comment]
Myt repo is a bit messy and I can't find the right posenet to the depthnet I uploaded on the drive. Sorry about that, shame on me ! :sob: , I'll rerun a training and upload the depthnet and posent together.

**Comment by sunnyHelen** [comment]
That's ok. It seems that the best models for disparity and pose are from different epochs. Can we use similar way to keep a best model for pose?

**Comment by jinfagang** [comment]
@ClementPinard  Yeah I just download this repo and find a little while where the model goes... Does there any STOA depth estimation methods for now? (both accuracy and speed) From your result gif image, the depth estimated seems  blurry

**Comment by ClementPinard** [comment]
Current state of the art (with available code) is probably struct2depth : https://sites.google.com/view/struct2depth

Other than that if you want to stay on pytorch, you might want to try simple enhancements like a more robust smooth loss (se here : https://torchgeometry.readthedocs.io/en/latest/losses.html#torchgeometry.losses.InverseDepthSmoothnessLoss )
you can also try to upscale warped image for a full resolution photometric loss (as explained by godard et al. here https://arxiv.org/abs/1806.01260 )

Finally, SSIM might be an option to have a more robust photometric loss (see here https://torchgeometry.readthedocs.io/en/latest/losses.html#torchgeometry.losses.SSIM )

To keep a clean basis, I chose to only use what the original author used, but it wouldn't be hard to implement these. Especially when pytorch geometry already does it for you

**Comment by sunnyHelen** [comment]
Hi, Clement. You mentioned the result will be better if pretrained on ImageNet. I want to know if the best model you provide here is pretrained on ImageNet. If not, which one will be better as a pretrained model?

**Comment by jinfagang** [comment]
@ClementPinard  thanks for your information, I think struct2depth is built on top of instance segmentation which may be slow and not clean. And I can not get it to train since it needs segmentation ground truth data but there were no material guide to train on cityscapes with raw data

**Comment by ClementPinard** [comment]
I don't recall mentioning imageNet as a pretraining ...

There is no pretraining involved here. Originally, there was pretraining on CityScapes, but ultimately, data augmentation of KITTI did a better job. see [here](https://github.com/tinghuiz/SfMLearner#notes)

**Comment by sunnyHelen** [comment]
Oh, I just see it in your paper here https://arxiv.org/abs/1806.01260.
In addition,  there are still many places I don't understand, especially the operation about pose. Could you please tell me what's the meaning of the following code? Why to do this?
![image](https://user-images.githubusercontent.com/28101548/55946720-89e45780-5c90-11e9-9b6d-728a927aec13.png)

**Comment by ClementPinard** [comment]
Haha, we have similar names, but this is not me! This is Clement godard's paper !
He's Mr haricot on github : https://github.com/mrharicot  sadly, he didn't opensource this paper.
The imagenet pretraining is probably anecdotical here, it's mainly for the encoder, that they used for both posenet and dispnet. This is not what we do here (with a dedicated network for each task), and I honestly think this wouldn't change anything in the training.

And as to why we do this little snippet here, is that we want the pose sequence to be relative to first pose. As such, the first pose will be identity and the other poses will be the difference of pose from the first one.

It's more or less the same thing that is done in original author code [here](https://github.com/tinghuiz/SfMLearner/blob/master/kitti_eval/pose_evaluation_utils.py#L371)

**Comment by sunnyHelen** [comment]
Oh, sorry. I made a stupid mistake. 
I'm trying to capture the best result for pose. It seems the best model it keeps now is just for depth, right? I want to imitate you to find a best model for pose, so I write the following codes. But It seems it doesn't work. The best model it kept is not the best . Could you please help me to find the error?
![image](https://user-images.githubusercontent.com/28101548/55949521-23face80-5c96-11e9-8422-5ae1ffae7184.png)
![image](https://user-images.githubusercontent.com/28101548/55949560-396ff880-5c96-11e9-999f-5474f953885f.png)
![image](https://user-images.githubusercontent.com/28101548/55949581-45f45100-5c96-11e9-90f4-1e89b34c8eb1.png)

**Comment by ClementPinard** [comment]
It would have been better to post the snippet in a gist, or in a github project forked from the original.

What do you mean the best model kept is not the best ? What is the function `compute_pose_errors1` ?

Apart from that I don't see big errors.

**Comment by sunnyHelen** [comment]
I mean I test the best model kept for pose, but it has worse result than the checkpoint one.
compute_pose_errors1 is same as the compute_pose_errors function in test_pose.py, except the way of return value.
![image](https://user-images.githubusercontent.com/28101548/55957763-747c2700-5caa-11e9-9833-da5d40f68def.png)

**Comment by ClementPinard** [comment]
The model that optimizes a validation set is not necessarily the one that optimizes the test set.
I see that in you validation step, you only test the first element of the batch of poses, which might worsen the metric used.

All in all, I see that the pose network has much less overfitting problem, so I would be surprised for example if the best model was in the first epochs for example. If that's the case, I would search for a bug then.

**Comment by sunnyHelen** [comment]
Thank you for your help. I'll think it over. In my experiments, the pose test result of kept best model is often worse than the checkpoint model. I don't know how to find the best model for pose test.

**Comment by sunnyHelen** [comment]
Hi, Clement. I have a question again. I want to know how to get the figures in those papers. 
![image](https://user-images.githubusercontent.com/28101548/57117924-ef64c900-6da2-11e9-8d5f-560c2245ba07.png)
![image](https://user-images.githubusercontent.com/28101548/57117948-1ae7b380-6da3-11e9-94a9-8cea8751008f.png)
I change the run_inference code to read the .npy files. But I got this.
![image](https://user-images.githubusercontent.com/28101548/57117965-3ce13600-6da3-11e9-933e-c154a1a46151.png)
![image](https://user-images.githubusercontent.com/28101548/57117975-54202380-6da3-11e9-9f1b-dbf4a4ebf5bb.png)
![image](https://user-images.githubusercontent.com/28101548/57118006-9ea1a000-6da3-11e9-8637-24ee4e98338f.png)
![image](https://user-images.githubusercontent.com/28101548/57118000-8e89c080-6da3-11e9-8b7c-6f3755673633.png)

**Comment by ClementPinard** [comment]
Have a look at https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.LinearNDInterpolator.html#scipy.interpolate.LinearNDInterpolator

here is a rough example of how it works from clement godard's repo : https://github.com/mrharicot/monodepth/blob/master/utils/evaluation_utils.py#L115
